### PR TITLE
Add CHIP stack module.

### PR DESCRIPTION
**Pull request by kghost**
#### Problem
CHIP stack initialization and shutdown are spread around. There are lots of duplicated code in all test cases and applications.

#### Change overview
This PR add a centralized class to initialize and shutdown all chip stack singletons.

#### Testing
This PR is a refactor, unit tests are used for verify.

I have manually tested echo requester/responder and im requester/responder. All verified are working.

### CODE REVIEW

File: src/stack/Stack.h [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+    CHIP_ERROR ResetTransport() { return mTransport.Init(mInetLayer.Get()); }
+
+    TransportConfig & GetTransportConfig() { return mTransport; }
+
+    NodeId GetLocalNodeId() { return mLocalDeviceId; }
+    System::Layer & GetSystemLayer() { return mSystemLayer.Get(); }
+    Inet::InetLayer & GetInetLayer() { return mInetLayer.Get(); }
+    BleLayerConfig & GetBleLayerConfig() { return mBleLayer; }
+    Ble::BleLayer * GetBleLayer() { return mBleLayer.Get(); }
```
Comment (andy31415) [inline src/stack/Stack.h:241]: A comment that these are nullable would be useful

File: src/lib/core/CHIPConfig.h [hunk 2]
```diff
@@ -86,6 +86,19 @@
...
+ *  @def CHIP_CONFIG_CHIP_IS_SINGLETON
+ *
+ *  @brief
+ *    This defines whether chip is singleton. If chip is singleton, all
+ *    pointers to chip components like SecureSessionMgr, ExchangeManager, etc
+ *    can be omitted, which saves memory on small platforms. For some situation
+ *    where there can be multiple chip instances, define this to 0.
+ */
+#ifndef CHIP_CONFIG_CHIP_IS_SINGLETON
+#define CHIP_CONFIG_CHIP_IS_SINGLETON 1
```
Comment (andy31415) [inline src/lib/core/CHIPConfig.h:99]: how do we test functiohnality when this is 0 or 1? Do we have separate builds validating that both versions work correctly?

File: src/stack/Stack.h [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+};
+
+class DefaultBleLayer : BleLayerConfiguration
+{
+public:
+    template <typename T>
+    CHIP_ERROR Init(T * stack)
+    {
+        return CHIP_NO_ERROR;
+    }
```
Comment (andy31415) [inline src/stack/Stack.h:136]: sholuld we have a Shutdown here as well so that we do not need to go through BLE nullptr checks during shutdown?

File: src/stack/Stack.h [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    {
+        return mTransportManager.Init(
+#if INET_CONFIG_ENABLE_IPV4
+            Transport::UdpListenParameters(&inetLayer).SetAddressType(Inet::kIPAddressType_IPv4).SetListenPort(mPort),
+#endif
+            Transport::UdpListenParameters(&inetLayer).SetAddressType(Inet::kIPAddressType_IPv6).SetListenPort(mPort));
+    }
+
+    TransportMgrBase & Get() { return mTransportManager; }
+    void SetListenPort(uint16_t port) { mPort = port; }
```
Comment (andy31415) [inline src/stack/Stack.h:169]: Since this is used during init, why are we storing this as a member? Could we pass this as part of the init arguments instead?

Comment (andy31415) [reply src/stack/Stack.h:169]: As a user, it sounds like I may want some 'StackInitParameters' to define how the stack is brought up. The listening port seems to be part of that.

I would rather have `gStack.Init(Args().SetListenPort(1234))` or maybe:

```cpp
StackParameters params;
params.SetListenPort(1234);

gStack.init(params)

```

than:
```cpp
gStack.GetTransportConfig().SetListenPort(1234);
gStack.Init();
```

because the latter relies on ordering and how arguments are passed around is not immediately clear.

File: src/stack/Traits.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+template <template <typename> class Cond, typename Default, typename T, typename... Ts>
+struct first_if_any_or_default<Cond, Default, T, Ts...>
+{
+private:
+    using next = first_if_any_or_default<Cond, Default, Ts...>;
+
+public:
+    using type                  = typename std::conditional<Cond<T>::value, T, typename next::type>::type;
+    static constexpr bool found = Cond<T>::value || next::found;
```
Comment (Damian-Nordic) [inline src/stack/Traits.h:42]: Is it used anywhere?

File: src/stack/Traits.h [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+};
+
+template <class, template <class, class...> class>
+struct is_instance : public std::false_type
```
Comment (Damian-Nordic) [inline src/stack/Traits.h:46]: Is it used?

**Review by jmartinez-silabs — APPROVED** [review]

**Comment by mspang** [comment]
This PR adds more code than it removes...

**Comment by mspang** [comment]
> This PR adds more code than it removes...

I'm also not convinced it's a step forward in maintainability. Now everyone will have to wrap their head around all of the metaprogramming to understand how to build a ChipStack. We should keep it simple.

**Comment by kghost** [comment]
@mspang 

I have tried my best to keep it simple. It is very hard to do compile time configuration, without introduce runtime overheads.


Althought this PR is adding more code than it removes, but most code is the stack itself. Lots of codes removed from application side, which will definitely reduce duplicated code of all chip vendors.

```
 src/stack/BUILD.gn                               |  33
 src/stack/Stack.h                                | 295
 src/stack/Traits.h                               |  43
```

Another problem of the old code is that initialization and shutdown calls are spread everywhere, makes it very difficult to solve issue #6931

**Comment by mspang** [comment]
> @mspang
> 
> I have tried my best to keep it simple. It is very hard to do compile time configuration, without introduce runtime overheads.
> 
> Althought this PR is adding more code than it removes, but most code is the stack itself. Lots of codes removed from application side, which will definitely reduce duplicated code of all chip vendors.
> 
> ```
>  src/stack/BUILD.gn                               |  33
>  src/stack/Stack.h                                | 295
>  src/stack/Traits.h                               |  43
> ```
> 
> Another problem of the old code is that initialization and shutdown calls are spread everywhere, makes it very difficult to solve issue #6931

For #6931 we should just merge the main thread and events thread together.

**Comment by kghost** [comment]
> we should just merge the main thread and IO thread together

The problem is that there are multiple singletons need to be shutdown by application. For example, assume application need to initialize and shutdown `ExchangeManager` and `SecureSessoinManager`, one could write like this:

```C++
gSecureSessoinManager.AsyncInit();
gSecureSessoinManager.WaitInitComplete();
gExchangeManager.AsyncInit();
gExchangeManager.WaitInitComplete();

...
... do works
...

gExchangeManager.AsyncShutdown();
gExchangeManager.WaitShutdownComplete();
gSecureSessoinManager.AsyncShutdown();
gSecureSessoinManager.WaitShutdownComplete();
```

Currently there are around 6-7 singletons to initialize and shutdown. It will take enormous work to implement. With help of this PR, one could write:

```
gStack.AsyncInit();
gStack.WaitInitComplete();

...
... do works
...

gStack.AsyncShutdown();
gStack.WaitShutdownComplete();
```

**Review by woody-apple — CHANGES_REQUESTED** [review]
Per the updated template, can you update the PR here?

```
#### Problem
What is being fixed?  Examples:
* Fix crash on startup
* Fixes #12345 12345 Frobnozzle is leaky (exactly like that, so GitHub will auto-close the issue).

#### Change overview
What's in this PR

#### Testing
How was this tested? (at least one bullet point required)
    • If unit tests were added, how do they cover this issue?
    • If unit tests existed, how were they fixed/modified to prevent this in future?
    • If integration tests were added, how do they verify this change?
    • If manually tested, what platforms controller and device platforms were manually tested, and how?
    • If no testing is required, why not?
```

**Review by bzbarsky-apple — COMMENTED** [review]
Why is removing all that code from `src/test_driver/esp32/main/main_app.cpp` ok?

File: src/stack/Stack.h [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+#include <stack/Stack.h>
+#include <stack/Traits.h>
+
+#include <messaging/ExchangeMgr.h>
+#include <protocols/secure_channel/MessageCounterManager.h>
+#include <transport/SecureSessionMgr.h>
+#include <transport/TransportMgr.h>
+#include <transport/raw/UDP.h>
+
+#if CONFIG_DEVICE_LAYER
```
[the remark this answers is not in the document]
Comment (andy31415) [reply src/stack/Stack.h:29]: GN defines it as:

```
buildconfig_header("system_buildconfig") {
  header = "SystemBuildConfig.h"
  header_dir = "system"
```

which is the header you require. You are currently inheriting it indirectly by some other header including it, however I believe this is dangerous as the ifdef then could silently appear/disappear.

File: src/test_driver/esp32/main/main_app.cpp [hunk 8]
```diff
@@ -81,52 +74,6 @@ extern "C" void app_main()
...
-        exit(err);
-    }
-    err = esp_event_handler_register(IP_EVENT, ESP_EVENT_ANY_ID, PlatformManagerImpl::HandleESPSystemEvent, NULL);
-    if (err != CHIP_NO_ERROR)
-    {
-        ESP_LOGE(TAG, "esp_event_handler_register() failed for IP_EVENT: %s", ErrorStr(err));
-        exit(err);
-    }
-
-    err = Crypto::add_entropy_source(app_entropy_source, NULL, 16);
```
Comment (andy31415) [inline src/test_driver/esp32/main/main_app.cpp:123]: crypto tests sound as if they may want an entropy source. Event loop also seems potentially important.

I would understand 'esp tests do not use wifi' because qemu, however this should be documented and probably a separate PR from the stack module unless it interferes with that...and if it does please explain why.

Comment (kghost) [reply src/test_driver/esp32/main/main_app.cpp:123]: This are duplicated in `PlatformManagerImpl::_InitChipStack`

https://github.com/project-chip/connectedhomeip/blob/82a8544c3a7df2e59e16f30fb7393df22a654a87/src/platform/ESP32/PlatformManagerImpl.cpp#L56

File: examples/shell/shell_common/include/Globals.h [hunk 9]
```diff
@@ -18,27 +18,59 @@
...
-constexpr size_t kMaxTcpActiveConnectionCount = 4;
-constexpr size_t kMaxTcpPendingPackets        = 4;
+
+class TransportConfigurationWithTcp : chip::TransportConfiguration
+{
+public:
+    static constexpr size_t kMaxTcpActiveConnectionCount = 4;
+    static constexpr size_t kMaxTcpPendingPackets        = 4;
+
+    using transport = chip::TransportMgr<
```
Comment (Damian-Nordic) [inline examples/shell/shell_common/include/Globals.h:36]: nit: I think we generally start type names with a capital letter. Change to "Transport"?

File: src/app/server/Server.cpp [hunk 10]
```diff
@@ -97,7 +98,77 @@ class ServerStorageDelegate : public PersistentStorageDelegate
...
-ServerStorageDelegate gServerStorage;
+class ServerStorageConfig : public StorageConfiguration
+{
+public:
+    CHIP_ERROR Init(const StackParameters & parameters)
+    {
+        CHIP_ERROR err = CHIP_NO_ERROR;
+
+#if CHIP_DEVICE_LAYER_TARGET_DARWIN
+        err = PersistedStorage::KeyValueStoreMgrImpl().Init("chip.store");
```
Comment (Damian-Nordic) [inline src/app/server/Server.cpp:109]: idea: Should we add `Init` method to the common `KeyValueStoreMgr` interface, add some "storageLocation" parameter to `StackParameters` and unify this initialization code by implementing `DefaultStorageConfiguration` when the device layer is present?

File: src/stack/Stack.h [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <stack/Stack.h>
```
Comment (Damian-Nordic) [inline src/stack/Stack.h:20]: self-reference?

File: src/stack/Stack.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (Damian-Nordic) [reply src/stack/Stack.h:29]: I think you can just include `ChipDeviceLayer.h` (unconditionally) which includes all necessary configs and `PlatformManager.h`.

File: src/stack/Stack.h [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+    using InetLayerConfig = typename first_if_any_or_default<IsInetLayerConfiguration, DefaultInetLayer, Configurations...>::type;
+    using StorageConfig   = typename first_if_any_or_default<IsStorageConfiguration, DefaultStorage, Configurations...>::type;
+    using BleLayerConfig  = typename first_if_any_or_default<IsBleLayerConfiguration, DefaultBleLayer, Configurations...>::type;
+    using TransportConfig = typename first_if_any_or_default<IsTransportConfiguration, DefaultTransport, Configurations...>::type;
+
+public:
+    Stack(NodeId localDeviceId) : mLocalDeviceId(localDeviceId) {}
+    virtual ~Stack() {}
+
+    // TODO(#6931): CHIP initialization should be posted to chip thread, in order to avoid racing problems
```
Comment (Damian-Nordic) [inline src/stack/Stack.h:221]: Is this true? Isn't this code supposed to be called before the chip thread is started?

File: src/stack/Stack.h [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+    CHIP_ERROR Init(const StackParameters & parameters)
+    {
+        ReturnErrorOnFailure(Platform::MemoryInit());
+#if CONFIG_DEVICE_LAYER
+        // Initialize the CHIP stack.
+        ReturnErrorOnFailure(DeviceLayer::PlatformMgr().InitChipStack());
```
Comment (Damian-Nordic) [inline src/stack/Stack.h:227]: FYI: For some reason, currently `ThreadStackMgr().InitThreadStack();` is also needed for Thread devices, so it either should be started here or incorporated into `InitChipStack()` when `CHIP_DEVICE_CONFIG_ENABLE_THREAD`

File: src/app/server/Server.cpp (hunk 10 above)
Comment (kghost) [reply src/app/server/Server.cpp:109]: This is a good suggestion, but this PR is only refactoring current code, I'd like it to be changed as less as possible.

File: src/stack/Stack.h (hunk 12 above)
Comment (kghost) [reply src/stack/Stack.h:221]: IMHO, the initialization will split to 3 parts:

 1. Initialize the platform: run in main thread, do hardware initialization
 2. Create the chip thread
 3. Initialize the chip components: run in chip thread

I'll work on it in the following PRs. Use issue #6931 to track the progress.

File: src/stack/Stack.h (hunk 13 above)
Comment (kghost) [reply src/stack/Stack.h:227]: I saw that `ThreadStackMgr().InitThreadStack()` is called after `InitChipStack()` for platforms with thread support. Maybe we should separate hardware initialization and chip stack initialization, and leave `InitChipStack()` outside chip stack.

Comment (Damian-Nordic) [reply src/stack/Stack.h:227]: I'm not sure why ThreadStackManager can't be initialized by InitChipStack (which initializes other platform components). Maybe it's because ThreadStackManager can post events to the CHIP thread and the intention was not to initialize ThreadStackManager before the CHIP thread is started. Anyway, in case you plan to move the CHIP thread creation to the CHIP stack it would make sense to initialize ThreadStackManager in the comon code, too.

**Comment by gerickson** [comment]
To the comment, "2. Create the CHIP thread", per #6561, do we absolutely need a CHIP thread? Ideally, in the simples desktop or mobile app or in the simplest RTOS system, we could have a single thread that provides the animating context for a number of different sources of work.

**Comment by Damian-Nordic** [comment]
@kghost Will the work be continued at some point?
